File: Exemplo-Java/Composite/src/br/edu/ifes/poo2/composite/app/TesteComposite.java
package br.edu.ifes.poo2.composite.app;

import br.edu.ifes.poo2.composite.pattern.DocumentoComposto;
import br.edu.ifes.poo2.composite.pattern.DocumentoSimples;


public class TesteComposite {
    
    public static void main(String[] args) {
        
    /* Montagem de uma árvore de composites */
        
    // Documento Raiz:
    DocumentoComposto raiz = new DocumentoComposto("c:/Raiz");
    
    //Dois filhos compostos do documento raiz:
    DocumentoComposto filhocomp1 = new DocumentoComposto("/FilhoComposto1");
    DocumentoComposto filhocomp2 = new DocumentoComposto("/FilhoComposto2");
    raiz.adicionarfilho(filhocomp1);
    raiz.adicionarfilho(filhocomp2);
    
    //Dois documentos simples, que serão filhos do documento composto
    //filhocomp1:
    DocumentoSimples ds1 = new DocumentoSimples("/Simples1");
    DocumentoSimples ds2 = new DocumentoSimples("/Simples2");
    filhocomp1.adicionarfilho(ds1);
    filhocomp1.adicionarfilho(ds2);
    
    //Documento composto que será "neto" do documento raiz, filho de
    //filhocomp2:
    DocumentoComposto netocomp = new DocumentoComposto("/netocomp");
    filhocomp2.adicionarfilho(netocomp);
    
    //Documentos simples, que serão filhos de netocomp
    DocumentoSimples ds3 = new DocumentoSimples("/Simples3");
    DocumentoSimples ds4 = new DocumentoSimples("/Simples4");
    netocomp.adicionarfilho(ds3);
    netocomp.adicionarfilho(ds4);
    
    //Imprimir a árvore de documentos a partir do documento raiz:
    raiz.imprimirURL("");
    
    }
}

File: Exemplo-Scala/Composite/src/composite/TesteComposite.scala
package composite

import java.util.ArrayList
import java.lang.ref.ReferenceQueue.Null

abstract class Documento (URL: String){
    
	def url: String = URL
  
    def imprimirURL(edent: String)
    
    def adicionarfilho(documento: Documento)
    
    def removerfilho(documento: Documento)
    
    def retornarfilho(index: Int): Documento
    
    def numfilhos: Int
}

class DocumentoComposto(URL: String) extends Documento (URL: String){
    
    var docsFilhos: ArrayList[Documento] = new ArrayList
	
    override def url: String = URL
    
    override def imprimirURL(edent: String){
      print(edent)
      print(url)
      print("\n")
      var index: Int = 0
      while(index < docsFilhos.size()){
        docsFilhos.get(index)imprimirURL(edent.concat("-"))
        index = index + 1
      }
	}

	override def adicionarfilho(documento: Documento){
		docsFilhos.add(documento);
    }

	override def removerfilho(documento: Documento){
	  docsFilhos.remove(documento)
	}
	
    override def retornarfilho(index: Int): Documento = docsFilhos.get(index)
    
    override def numfilhos: Int = docsFilhos.size()
}
 
class DocumentoSimples(URL: String) extends Documento (URL: String){
	
    override def url: String = URL
    
    override def imprimirURL(edent: String){
      print(edent)
      print(url)
      print("\n")
	}

	override def adicionarfilho(documento: Documento){}
	
	override def removerfilho(documento: Documento){}

    override def retornarfilho(index: Int): Documento = null
    
    override def numfilhos: Int = 0
}


object TesteComposite {
  def main(args: Array[String]) = {
    var raiz: DocumentoComposto = new DocumentoComposto("c:/Raiz");
    
    var filhocomp1: DocumentoComposto  = new DocumentoComposto("/FilhoComposto1");
    var filhocomp2: DocumentoComposto  = new DocumentoComposto("/FilhoComposto2");
    raiz.adicionarfilho(filhocomp1);
    raiz.adicionarfilho(filhocomp2);
    
    var ds1: DocumentoSimples = new DocumentoSimples("/Simples1");
    var ds2: DocumentoSimples = new DocumentoSimples("/Simples2");
    filhocomp1.adicionarfilho(ds1);
    filhocomp1.adicionarfilho(ds2);
 
    var netocomp: DocumentoComposto  = new DocumentoComposto("/netocomp");
    filhocomp2.adicionarfilho(netocomp);
    
    var ds3: DocumentoSimples = new DocumentoSimples("/Simples3");
    var ds4: DocumentoSimples = new DocumentoSimples("/Simples4");
    netocomp.adicionarfilho(ds3);
    netocomp.adicionarfilho(ds4);
   
    raiz.imprimirURL("");
    
  }
}
